engine/gamesys/src/gamesys/scripts/script_buffer.cpp

```diff
@@ -879,6 +880,209 @@ namespace dmGameSystem
+                    break;
+                    case dmBuffer::VALUE_TYPE_UINT32:
+                        values = (void*) luaTableToArray<uint32_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_UINT64:
+                        values = (void*) luaTableToArray<uint64_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT8:
+                        values = (void*) luaTableToArray<int8_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT16:
+                        values = (void*) luaTableToArray<int16_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT32:
+                        values = (void*) luaTableToArray<int32_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT64:
+                        values = (void*) luaTableToArray<int64_t>(L, count, valueType);
+                    break;
+                    default:
+                        return DM_LUA_ERROR("%s.%s invalid value type supplied: %ld.", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, valueTypeRaw);
+                }
+
+                dmBuffer::Result r = dmBuffer::SetMetaData(hbuffer, entry_name, (void*) values, count, valueType);
+                free(values);
+
+                if( r != dmBuffer::RESULT_OK )
+                {
+                    return DM_LUA_ERROR("%s.%s: cannot set metadata for buffer: %s", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, dmBuffer::GetResultString(r));
+                }
+            } else {
+                return DM_LUA_ERROR("%s.%s: invalid metadata", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA);
+            }
+        } else
+            return DM_LUA_ERROR("%s.%s: invalid number of arguments", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA);
+
+        return 0;
+    }
+
+    // copies values from array to (empty) lua table on top of the stack
+    template <typename T>
+    static void arrayToLuaTable(lua_State* L, const T* values, uint32_t count, dmBuffer::ValueType valueType) {
+        if (valueType == dmBuffer::VALUE_TYPE_FLOAT32) {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushnumber(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        } else {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushinteger(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        }
+    }
+
+    // local metadata, numbertype  = buffer.get_metadata(self.new_buffer, "aabb")
+    static int GetMetadata(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 2); // return metadata table and value type
+        int top = lua_gettop(L);
+
+        if (top == 2)
```

Comment:
Same question as above about check number of values on the stack

Reply:
But if we decide to keep it I think I'd prefer to early out if there aren't enough arguments:

```
if (top < 2)
{
    return DM_LUA_ERROR("invalid number of arguments. Expected %d", 2);
}

// rest of code here

Reply:
ok, removing and waiting for final confirmation.

Reply:
I added a test case for that too (fewer arguments in get_metadata()). It works exactly as you said, resulting in a message like this:

DEBUG:SCRIPT: [string "src/gamesys/test/buffer/metadata.script"]:130: bad argument #2 to 'get_metadata' (hash or string expected expected, got no value)

Only trying to get the values of the arguments with lua_check* function is enough and much cleaner.